cherrypick: build,sql,cli: reveal the CockroachDB version number

knz opened the pull request:
Previous to this patch, the vtable `crdb_internal.node_build_info`,
the built-in SQL function `version()` and the command-line utilities
would only print the build tag as a version number.

This patch augments this display by prefixing it by the
`VersionPrefix()`, computed after the current branch.

- for the branch `release-1.0`, where the last commit in the branch is
  exactly a tag, the tag string is exactly `v1.0.6`.
- for the branch `release-1.1`, where there were some commits since the
  last tag, the tag string is something like
  `v1.1-beta.20170907-95-gdcc6ca4c3`.
- for the master branch, once we have at least one 1.2 alpha release,
  the tag string will be something like
  `v1.2-alpha.2017MMDD-NN-SHAAAAAAA`.

cc @cockroachdb/release

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18562)
<!-- Reviewable:end -->

bdarnell commented:
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/build/info.go, line 81 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18562#-KuLYzh8BKOENYXkVluy:-KuLYzh8BKOENYXkVluz:b-26g3dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7e450ca5967e59be594fc0d73705623fe421a22/pkg/build/info.go#L81)):*
> ```Go
> // VersionedTag prefixes the tag with VersionPrefix.
> func (b Info) VersionedTag() string {
> 	return VersionPrefix() + "-" + b.Tag
> ```

I don't think this is going to the right thing for release builds. In those builds, `b.Tag` is a tag and looks like `v1.0.6`, not a commit hash. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18562)*
<!-- Sent from Reviewable.io -->

knz commented:
I'll rebase with a different commit.

knz committed:
build,sql,cli: replace/fix the previous patch that revealed the version number

The previous patch was blindly concatenating the result of
`VersionPrefix()` with the build tag, to yield something like
`v1.1-abcdef`. However, as highlighted by @bdarnell, in the specific
case where the current tree is exactly a tagged version, then the
build tag becomes the (versioned) git tag, and the resulting visible
version number becomes `v1.1-v1.1`, which ... makes no sense.

This patch resolves this by removing the custom logic introduced by
the previous patch entirely, and using the output of `git describe`
instead of `git describe --exact-match` as buildinfo tag.

This gives us:

- for the branch `release-1.0`, where the last commit in the branch is
  exactly a tag, the tag string is exactly `v1.0.6`.
- for the branch `release-1.1`, where there were some commits since the
  last tag, the tag string is something like
  `v1.1-beta.20170907-95-gdcc6ca4c3`.
- for the master branch, once we have at least one 1.2 alpha release,
  the tag string will be something like
  `v1.2-alpha.2017MMDD-NN-SHAAAAAAA`.

knz commented:
Using the new fix, PTAL.
